File: packages/core/src/core/renderer.ts
/**
 * Normalizes raw markdown output into clean, well-formatted text.
 *
 * Performs the following transformations:
 * - Normalizes line endings to `\n`
 * - Collapses whitespace-only lines into empty lines
 * - Collapses 3+ consecutive newlines into exactly 2
 * - Trims trailing whitespace from each line (preserving intentional `  ` line breaks)
 * - Ensures the output ends with a single newline
 *
 * @param raw - The raw markdown string produced by the walker.
 * @returns The cleaned and normalized markdown string.
 * @internal
 */
export function render(raw: string): string {
    return (
        raw
            .replaceAll(/\r\n?/g, '\n')
            // Collapse whitespace-only lines into empty lines
            .replaceAll(/\n[ \t]+\n/g, '\n\n')
            // Collapse 3+ consecutive newlines into 2
            .replaceAll(/\n{3,}/g, '\n\n')
            .split('\n')
            .map(line => {
                if (/\S {2}$/.test(line)) {
                    return line.replace(/\s+$/, '') + '  ';
                }
                return line.trimEnd();
            })
            .join('\n')
            .trim() + '\n'
    );
}

File: python/src/markdown_for_agents/core/renderer.py
from __future__ import annotations

import re


def render(raw: str) -> str:
    """Normalize raw markdown output into clean, well-formatted text."""
    # Normalize line endings to \n
    result = re.sub(r"\r\n?", "\n", raw)
    # Collapse whitespace-only lines into empty lines
    result = re.sub(r"\n[ \t]+\n", "\n\n", result)
    # Collapse 3+ consecutive newlines into 2
    result = re.sub(r"\n{3,}", "\n\n", result)
    # Trim trailing whitespace from each line, preserving intentional double-space
    lines = result.split("\n")
    processed: list[str] = []
    for line in lines:
        if re.search(r"\S {2}$", line):
            processed.append(line.rstrip() + "  ")
        else:
            processed.append(line.rstrip())
    result = "\n".join(processed)
    return result.strip() + "\n"
